test/functional/p2p_filter.py

```diff
@@ -104,6 +105,18 @@ def run_test(self):
             txid = self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 7)
             filter_node.wait_for_tx(txid)
 
+        self.log.info('Check that request for filtered blocks is ignored if no filter is set')
+        filter_node.merkleblock_received = False
+        self.nodes[0].generatetoaddress(1, self.nodes[0].getnewaddress())
+        filter_node.sync_with_ping()
+        filter_node.sync_with_ping()
```

Comment:
Why is this calling `sync_with_ping`? Shouldn't it send a getdata(blockhash, MSG_FILTERED_BLOCK)?

Reply:
Sending of `getdata` happens in the `on_inv()` callback method, i.e. as reply to an received `inv` message from the node. The `sync_with_ping`s are there to have some implicit delay to ensure there was enough time for the `inv`/`getdata` exchange and a (in this case, fictive) `merkleblock` message to arrive, so that the assert below has a value. (If you remember, we discussed in PR https://github.com/bitcoin/bitcoin/pull/18334 about this multiple-ping approach; I was [sceptical](https://github.com/bitcoin/bitcoin/pull/18334#issuecomment-605617660) but [your explanation](https://github.com/bitcoin/bitcoin/pull/18334#issuecomment-605624202) led me to the conclusion that there is no better way of doing this.)

Reply:
Oh, I didn't realize that inv's and getdata's are already sent back and forth. Maybe clarify this with an `assert_debug_log`?

Reply:
That's a good idea. I haven't used `assert_debug_log` before but from what I can see it _waits_ for the expected messages, i.e. after such an debug-log-assert block with the expected messages of `'sending inv'` and `'received getdata'` we have the guarantee that the node has already received the `getdata` request and one following `sync_with_ping` should be enough. :tada:

Reply:
Yeah, it happens to wait for it to appear, but we shouldn't rely on that when possible. This has been added only as an "emergency measure" for `rpc_setban.py`. ;)